packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts

```diff
@@ -197,43 +166,64 @@ export const useGridSelection = (apiRef: GridApiRef): void => {
   useGridApiMethod(apiRef, selectionApi, 'GridSelectionApi');
 
   React.useEffect(() => {
+    apiRef.current.registerControlState<GridSelectionModel>({
+      stateId: 'selectionModel',
+      propModel: props.selectionModel,
+      propOnChange: props.onSelectionModelChange,
+      stateSelector: gridSelectionStateSelector,
+      onChangeCallback: (model: GridSelectionModel) => {
+        apiRef.current.publishEvent(GRID_SELECTION_CHANGE, model);
+      },
+    });
+  }, [apiRef, props.onSelectionModelChange, props.selectionModel]);
+
+  React.useEffect(() => {
+    // Rows changed
     setGridState((state) => {
-      const newSelectionState = { ...state.selection };
+      const newSelectionState = [...state.selection];
+      const selectionLookup = selectedIdsLookupSelector(state);
+
       let hasChanged = false;
-      Object.keys(newSelectionState).forEach((id: GridRowId) => {
+      newSelectionState.forEach((id: GridRowId) => {
         if (!rowsLookup[id]) {
-          delete newSelectionState[id];
+          delete selectionLookup[id];
           hasChanged = true;
         }
       });
       if (hasChanged) {
-        return { ...state, selection: newSelectionState };
+        return { ...state, selection: Object.values(selectionLookup) };
       }
       return state;
     });
     forceUpdate();
   }, [rowsLookup, apiRef, setGridState, forceUpdate]);
 
   React.useEffect(() => {
-    const currentModel = Object.values(apiRef.current.getState().selection);
-    if (!isDeepEqual(currentModel, selectionModel)) {
-      apiRef.current.setSelectionModel(selectionModel || []);
+    // prop selectionModel changed
+    if (props.selectionModel === undefined) {
+      return;
+    }
+    const currentModel = apiRef.current.getState().selection;
+    if (currentModel !== props.selectionModel) {
+      setGridState((state) => ({ ...state, selection: props.selectionModel || [] }));
```

Comment:
By not using `setSelectionModel`, the user can hijack the grid and create a DataGrid with multiple selection. Look https://codesandbox.io/s/material-demo-forked-essgk?file=/demo.js